File: banshee/banshee.pl
#!/usr/bin/perl -w



# XChat-Banshee-Plugin v1.4
#
# Author: Will Farrington <wcfarrington@gmail.com>

#

# This program is free software; you can redistribute it and/or modify

# it under the terms of the GNU General Public License as published by

# the Free Software Foundation; either version 2 of the License, or

# (at your option) any later version.

#

# This program is distributed in the hope that it will be useful,

# but WITHOUT ANY WARRANTY; without even the implied warranty of

# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the

# GNU General Public License for more details.


# This script allows the user to control Banshee and print playing media from Xchat. It was written over the course of several weeks. As such there are a few coding guidelines I ask if anyone wants to submit revisions to the code:

# Comments are required. Comments describing a subroutine should be prefaced with "##" and entirely capitalized, while comments inside a subroutine or not describing one should use "#" and be in traditional sentence form.

# Extra linebreaks are appreciated but not required.

# When the user submits ANY command, the script should offer in-Xchat feedback about that command.

# When at all possible, DBUS should be preferred over any sort of BASH command flag.


use Net::DBus;


# Register the plugin in Xchat.


$script_name = "XChat-Banshee Plugin";

$script_version = "1.4";

$script_description = "Integrates XChat with Banshee - GNOME Media Player.";



Xchat::register($script_name,$script_version,$script_description,\&unload);


# Let the user know the plugin has been loaded.


Xchat::print("Loaded \002".$script_name."\002:");

Xchat::print("Use \002/mhelp\002 to display a list of commands.");



# Register the commands.



Xchat::hook_command("mhelp",\&mhelp);

Xchat::hook_command("media",\&media);

Xchat::hook_command("next",\&next);

Xchat::hook_command("prev",\&prev);

Xchat::hook_command("play",\&play);

Xchat::hook_command("pause",\&pause);

Xchat::hook_command("rate",\&rate);



## HELP 



sub mhelp

{

# Display help screen.

Xchat::print("\002\037".$script_name." Help:\037\002");

Xchat::print(" \002About:\002");

Xchat::print(" * Author: Will Farrington [www.kalmwave.com] <wcfarrington@gmail.com>");

Xchat::print(" * Script Version: ".$script_version);

Xchat::print(" * XChat Version: ".Xchat::get_info(0));

Xchat::print(" \002Commands:\002");

Xchat::print(" * /mhelp - Display this screen.");

Xchat::print(" * /media - Display the current song playing to a channel.");

Xchat::print(" * /next - Skips to next song.");

Xchat::print(" * /prev - Skips to previous song.");

Xchat::print(" * /play - Starts playback.");

Xchat::print(" * /pause - Pauses playback.");

Xchat::print(" * /rate - Allows rating of currently playing song.");

}



# Grab the current Banshee version.



$banshee_version = `banshee --version`;

chop $banshee_version;


# Declare these variables now so that when we run the subroutine grab_song, it works like we want.

my ($artist, $track, $pos, $dur, $true_pos, $true_dur, $rating, $maxrating, $album, $rating2, $obj, $iface);
 

## TEST DBUS



sub test_dbus

{

	my $service = shift;

 

	my $bus = Net::DBus->session;

	my $obj = $bus->get_service("org.freedesktop.DBus");

	my $iface = $obj->get_object("/org/freedesktop/DBus", 

		"org.freedesktop.DBus");

 

	foreach my $item( @{$iface->ListNames()} )

	{

		if( $item eq $service )

		{

			return 1;

		}

	}

 

	return 0;

}



# Set DBUS vars for Banshee.



my($data, $server, $witem) = @_;

 

my $bus = Net::DBus->session;

 
## ESTABLISH IF BANSHEE IS RUNNING IN DBUS


sub dbus_stuff {
if(!&test_dbus("org.gnome.Banshee"))

{

	Xchat::print("Could not see Banshee in dbus.");
	return Xchat::EAT_PLUGIN;

}

# Set Banshee's "address" for DBUS.


$obj = $bus->get_service("org.gnome.Banshee");

$iface = $obj->get_object("/org/gnome/Banshee/Player",

		"org.gnome.Banshee.Core");
}



## GRAB SONG INFORMATION



sub grab_song

{
dbus_stuff();


# Get current playing song information.

$artist = $iface->GetPlayingArtist;

$track = $iface->GetPlayingTitle;

$album = $iface->GetPlayingAlbum;

$true_pos = $iface->GetPlayingPosition;

$true_dur = $iface->GetPlayingDuration;

$rating = $iface->GetPlayingRating;

$maxrating = $iface->GetMaxRating;

# If the song hasn't been rated, show "Not Rated" instead of "0/5".



if ($rating == 0) {

$rating2 = "Not Rated.";

} else {

$rating2 = $rating;

}


# Fix for Banshee duration and position.

chop $pos;

chop $dur;

$pos = sprintf("%d:%02d", int($true_pos / 60), $true_pos % 60);

$dur = sprintf("%d:%02d", int($true_dur / 60), $true_dur % 60);



return 0;

}



## PRINT PLAYING MEDIA



sub media

{

grab_song();


# Say the current playing song in the channel.


# If the song has no rating yet, show "Not Rated" instead.
	if ($rating == 0) {
		

Xchat::command("me is listening to 12$track01 by 13$artist01 on 10$album01 using $banshee_version Rated: $rating2 ($pos/$dur)");



} else { 



Xchat::command("me is listening to 12$track01 by 13$artist01 on 10$album01 using $banshee_version Rated: $rating2"."/"."$maxrating ($pos/$dur)"); 



}

return 1;

}


## SKIP TRACK


sub next

{
dbus_stuff();


# Skip to the next track.

$iface->Next;


# Echo in the window what we just did.
Xchat::print("Skipped to next track.");

return 1;

}


## SKIP TO LAST TRACK


sub prev

{
dbus_stuff();


# Skip to the previous track.

$iface->Previous;


# Echo in the window what we just did.
Xchat::print("Skipped to previous track.");

return 1;

}


## PLAY MEDIA


sub play

{
dbus_stuff();


# Start playback.

$iface->Play;


# Echo in the window what we just did.
Xchat::print("Started playback.");

return 1;

}



## PAUSE MEDIA


sub pause

{
dbus_stuff();


# Pause playback.

$iface->Pause;

# Echo in the window what we just did.

Xchat::print("Paused playback.");

return 1;

}


## RATE MEDIA


sub rate {

grab_song();
use Net::DBus qw(:typing);


# Rate the song. Notice the use of dbus_int32 to get a dbus-compliant integer as the result.
$newrating = dbus_int32($_[0][1]);

$iface->SetPlayingRating($newrating);

# Echo in the window what we just did.

Xchat::print("Rated $track by $artist a $_[0][1]/$maxrating.");

return 1;

}

File: xchat-gnome/banshee.py
#!/usr/bin/python

# Banshee Plugin for Xchat
# Copyright (C) 2007  Will Farrington <wcfarrington@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import xchat
import dbus

__module_name__ = "Banshee Plugin for Xchat"
__module_version__ = "1.0"
__module_description__ = "Announce music from and control Banshee Music Player."

bus = dbus.SessionBus()

class Dbus:
    def __init__(self):
        obj = bus.get_object('org.freedesktop.DBus', '/org/freedesktop/DBus')
        iface = dbus.Interface(obj, 'org.freedesktop.DBus')
        if 'org.gnome.Banshee' in iface.ListNames():
            banshee_obj = bus.get_object('org.gnome.Banshee', '/org/gnome/Banshee/Player')
            global banshee_iface
            banshee_iface = dbus.Interface(banshee_obj, 'org.gnome.Banshee.Core')
        else:
            xchat.prnt("\002Banshee is not registered with DBus. Please start Banshee.\002")
            return
    
    def grab_track_data(self):
        global track_title
        global track_artist
        global track_album
        global track_rating
        global track_pos
        global track_dur
        
        track_title = banshee_iface.GetPlayingTitle()
        track_artist = banshee_iface.GetPlayingArtist()
        track_album = banshee_iface.GetPlayingAlbum()
        track_rating = banshee_iface.GetPlayingRating()
        track_pos = banshee_iface.GetPlayingPosition()
        track_dur = banshee_iface.GetPlayingDuration()
        
        if track_rating == 0:
            track_rating = "Not Rated."
        else:
            track_rating = "%s/5" % track_rating
            
        track_pos = int(track_pos)
        track_dur = int(track_dur)
        track_pos = "%d:%02d" % (track_pos / 60, track_pos % 60)
        track_dur = "%d:%02d" % (track_dur / 60, track_dur % 60)

class Commands:
    def mabout(self, word, word_eol, nil):
        xchat.prnt("\002\037%s\037\002" % __module_name__)
        xchat.prnt(" \002About:\002")
        xchat.prnt(" * Author: Will Farrington <wcfarrington@gmail.com>")
        xchat.prnt(" * Script Version: %s" % __module_version__)
        xchat.prnt(" \002Help:\002")
        xchat.prnt(" * \037/media\037 - Displays the currently playing track")
        xchat.prnt(" * \037/prev\037 - Skips to the previous track")
        xchat.prnt(" * \037/stop\037 - Stops playback")
        xchat.prnt(" * \037/play\037 - Starts playback")    
        xchat.prnt(" * \037/pause\037 - Pauses playback")
        xchat.prnt(" * \037/next\037 - Skips to the next track")
        xchat.prnt(" * \037/rate n\037 - Rates the song n/5")
        
    def media(self, word, word_eol, nil):
        Dbus().grab_track_data()
        xchat.command("me is listening to: \"%s\" by %s on %s; Rated: %s (%s/%s)" % (track_title, track_artist, track_album, track_rating, track_pos, track_dur))
        
    def next(self, word, word_eol, nil):
        Dbus()
        banshee_iface.Next()
        xchat.prnt("Skipped to next track.")
        
    def prev(self, word, word_eol, nil):
        Dbus()
        banshee_iface.Previous()
        xchat.prnt("Skipped to previous track.")
    
    def play(self, word, word_eol, nil):
        Dbus()
        banshee_iface.Play()
        xchat.prnt("Started playback.")
    
    def pause(self, word, word_eol, nil):
        Dbus()
        banshee_iface.Pause()
        xchat.prnt("Paused playback.")
        
    def rate(self, word, word_eol, new_rating):
        Dbus().grab_track_data()
        new_rating = int(word[1])
        banshee_iface.SetPlayingRating(new_rating)
        xchat.prnt("Rated \"%s\" by %s a %s/5." % (track_title, track_artist, new_rating))

xchat.prnt("Loaded \002%s\002:" % __module_name__)
xchat.prnt("Use \002/mabout\002 to display a list of commands.")

mabout_hook = xchat.hook_command("mabout", Commands().mabout)
media_hook  = xchat.hook_command("media", Commands().media)
next_hook   = xchat.hook_command("next", Commands().next)
prev_hook   = xchat.hook_command("prev", Commands().prev)
play_hook   = xchat.hook_command("play", Commands().play)
pause_hook  = xchat.hook_command("pause", Commands().pause)
rate_hook   = xchat.hook_command("rate", Commands().rate)
        
def munload(nil):
    xchat.unhook(mabout_hook)
    xchat.unhook(media_hook)
    xchat.unhook(next_hook)
    xchat.unhook(prev_hook)
    xchat.unhook(play_hook)
    xchat.unhook(pause_hook)
    xchat.unhook(rate_hook)
    xchat.prnt("Thanks for using my script! =)")

xchat.hook_unload(munload)
